File: internal/services/notify/notify.go
// Package notify sends push notifications through a user's own Apprise
// channels when new hints appear, respecting per-channel severity and
// quiet hours.
//
//	every minute   new hints ≥ channel level, not sent yet, outside quiet hours → push
//
// Digest e-mails (Python's notify.digests) are not ported: that needs an
// SMTP outbound service that doesn't exist in Go yet.
package notify

import (
	"context"
	"database/sql"
	"errors"
	"strconv"
	"strings"
	"time"

	"dashboard/internal/crypto"
	"dashboard/internal/db"
	"dashboard/internal/enums"
	"dashboard/internal/i18n"
	"dashboard/internal/model"
	"dashboard/internal/outbound"
	data "dashboard/internal/repos/data"
	"dashboard/internal/repos/users"
	"dashboard/internal/services/access"
	"dashboard/internal/services/audit"
	"dashboard/internal/services/hints"
	"dashboard/internal/services/util"
	"dashboard/internal/settings"
)

const maxDigestLines = 10

var berlin = mustLoadBerlin()

func mustLoadBerlin() *time.Location {
	loc, err := time.LoadLocation("Europe/Berlin")
	if err != nil {
		return time.UTC
	}
	return loc
}

var (
	ErrDenied     = access.ErrDenied
	ErrInvalidURL = errors.New("notify: invalid apprise url")
	ErrBadTime    = errors.New("notify: bad time")
	ErrFailed     = errors.New("notify: delivery failed")
)

// ChannelView is one channel as shown to its owner (URL masked).
type ChannelView struct {
	ID          int64
	Name        string
	Hint        string // masked URL, e.g. "ntfy://…/topic"
	MinSeverity enums.Severity
	Enabled     bool
}

func mask(rawURL string) string {
	scheme, rest, ok := strings.Cut(rawURL, "://")
	if !ok {
		return scheme
	}
	if i := strings.LastIndex(rest, "/"); i >= 0 {
		return scheme + "://…/" + rest[i+1:]
	}
	return scheme + "://" + rest
}

func looksLikeApprise(rawURL string) bool {
	return strings.Contains(rawURL, "://")
}

// Channels lists a user's own notification channels.
func Channels(d *sql.DB, who *access.Principal) ([]ChannelView, error) {
	var out []ChannelView
	err := db.WithTx(d, func(tx *sql.Tx) error {
		chans, err := data.Channels(tx, who.UserID)
		if err != nil {
			return err
		}
		for _, c := range chans {
			url, err := crypto.Decrypt(c.URLEnc, crypto.PurposeNotify)
			if err != nil {
				return err
			}
			out = append(out, ChannelView{ID: c.ID, Name: c.Name, Hint: mask(url), MinSeverity: c.MinSeverity, Enabled: c.Enabled})
		}
		return nil
	})
	return out, err
}

// AddChannel adds a new apprise:// (or ntfy://, gotify://, ...) channel.
func AddChannel(d *sql.DB, who *access.Principal, name, rawURL string, level enums.Severity) error {
	rawURL = strings.TrimSpace(rawURL)
	if !looksLikeApprise(rawURL) {
		return ErrInvalidURL
	}
	label := strings.TrimSpace(name)
	if label == "" {
		label, _, _ = strings.Cut(rawURL, ":")
	}
	enc, err := crypto.Encrypt(rawURL, crypto.PurposeNotify, nil)
	if err != nil {
		return err
	}
	return db.WithTx(d, func(tx *sql.Tx) error {
		channel := &model.NotifyChannel{UserID: who.UserID, Name: label, URLEnc: enc, MinSeverity: level, Enabled: true}
		if err := data.AddChannel(tx, channel); err != nil {
			return err
		}
		return audit.Log(tx, &who.UserID, "notify.channel_added", label, "", nil)
	})
}

func own(q db.Queryer, who *access.Principal, channelID int64) (*model.NotifyChannel, error) {
	item, err := data.Channel(q, channelID)
	if err != nil {
		return nil, err
	}
	if item == nil || item.UserID != who.UserID {
		return nil, ErrDenied
	}
	return item, nil
}

// DeleteChannel removes one of the caller's own channels.
func DeleteChannel(d *sql.DB, who *access.Principal, channelID int64) error {
	return db.WithTx(d, func(tx *sql.Tx) error {
		if _, err := own(tx, who, channelID); err != nil {
			return err
		}
		return data.RemoveChannel(tx, channelID)
	})
}

// TestChannel sends a test push through one of the caller's own channels.
func TestChannel(ctx context.Context, d *sql.DB, cfg settings.Settings, who *access.Principal, channelID int64) error {
	var rawURL string
	err := db.WithTx(d, func(tx *sql.Tx) error {
		channel, err := own(tx, who, channelID)
		if err != nil {
			return err
		}
		rawURL, err = crypto.Decrypt(channel.URLEnc, crypto.PurposeNotify)
		return err
	})
	if err != nil {
		return err
	}
	title := i18n.T("notify.test_title", who.Locale, nil)
	body := i18n.T("notify.test_body", who.Locale, nil)
	if err := outbound.Send(ctx, cfg.AppriseAPIURL, rawURL, title, body); err != nil {
		return ErrFailed
	}
	return nil
}

// ── Quiet hours ──

// Prefs is a user's notification preferences.
type Prefs struct {
	QuietFrom, QuietTo string // "HH:MM", "" = no quiet hours
}

const quietKey = "quiet"

// GetPrefs reads a user's quiet-hours preference.
func GetPrefs(d *sql.DB, who *access.Principal) (Prefs, error) {
	var out Prefs
	err := db.WithTx(d, func(tx *sql.Tx) error {
		u, err := users.Get(tx, who.UserID)
		if err != nil || u == nil {
			return orNotFound(err)
		}
		quiet, _ := u.Prefs[quietKey].(map[string]any)
		out.QuietFrom, _ = quiet["from"].(string)
		out.QuietTo, _ = quiet["to"].(string)
		return nil
	})
	return out, err
}

// SavePrefs writes a user's quiet-hours preference.
func SavePrefs(d *sql.DB, who *access.Principal, p Prefs) error {
	for _, text := range []string{p.QuietFrom, p.QuietTo} {
		if text != "" && parseClock(text) == nil {
			return ErrBadTime
		}
	}
	return db.WithTx(d, func(tx *sql.Tx) error {
		u, err := users.Get(tx, who.UserID)
		if err != nil || u == nil {
			return orNotFound(err)
		}
		prefs := map[string]any{}
		for k, v := range u.Prefs {
			prefs[k] = v
		}
		prefs[quietKey] = map[string]any{"from": p.QuietFrom, "to": p.QuietTo}
		u.Prefs = prefs
		return users.Update(tx, u)
	})
}

func orNotFound(err error) error {
	if err != nil {
		return err
	}
	return util.ErrNotFound
}

// clock is minutes since midnight, for a quiet-hours comparison that
// doesn't need a full time.Time.
type clock int

func parseClock(text string) *clock {
	parts := strings.SplitN(text, ":", 2)
	if len(parts) != 2 {
		return nil
	}
	h, err1 := strconv.Atoi(parts[0])
	m, err2 := strconv.Atoi(parts[1])
	if err1 != nil || err2 != nil || h < 0 || h > 23 || m < 0 || m > 59 {
		return nil
	}
	c := clock(h*60 + m)
	return &c
}

// quietNow reports whether now falls in a user's quiet window (Europe/
// Berlin, may cross midnight: 22:00–07:00).
func quietNow(prefs map[string]any, now time.Time) bool {
	quiet, _ := prefs[quietKey].(map[string]any)
	from, _ := quiet["from"].(string)
	to, _ := quiet["to"].(string)
	start, end := parseClock(from), parseClock(to)
	if start == nil || end == nil {
		return false
	}
	local := now.In(berlin)
	current := clock(local.Hour()*60 + local.Minute())
	if *start <= *end {
		return *start <= current && current < *end
	}
	return current >= *start || current < *end
}

// ── Dispatch (job) ──

// Dispatch pushes new hints to every active user's enabled channels,
// skipping users currently in their quiet hours. Returns the number of
// pushes sent (one per channel per user, batched across hints).
func Dispatch(ctx context.Context, d *sql.DB, cfg settings.Settings) (int, error) {
	var people []*model.User
	err := db.WithTx(d, func(tx *sql.Tx) error {
		all, err := users.All(tx)
		if err != nil {
			return err
		}
		for _, u := range all {
			if u.IsActive {
				people = append(people, u)
			}
		}
		return nil
	})
	if err != nil {
		return 0, err
	}

	sent := 0
	now := time.Now().UTC()
	for _, u := range people {
		if quietNow(u.Prefs, now) {
			continue
		}
		n, err := dispatchUser(ctx, d, cfg, u.ID)
		if err != nil {
			return sent, err
		}
		sent += n
	}
	return sent, nil
}

func dispatchUser(ctx context.Context, d *sql.DB, cfg settings.Settings, userID int64) (int, error) {
	who, err := access.Load(d, userID)
	if err != nil {
		return 0, err
	}

	var chans []*model.NotifyChannel
	err = db.WithTx(d, func(tx *sql.Tx) error {
		found, err := data.Channels(tx, userID)
		if err != nil {
			return err
		}
		for _, c := range found {
			if c.Enabled {
				chans = append(chans, c)
			}
		}
		return nil
	})
	if err != nil || len(chans) == 0 {
		return 0, err
	}

	open, err := hints.Active(d, who, enums.SeverityInfo, nil, 0)
	if err != nil {
		return 0, err
	}

	var fresh []hints.View
	err = db.WithTx(d, func(tx *sql.Tx) error {
		for _, h := range open {
			sent, err := data.WasSent(tx, userID, h.ID)
			if err != nil {
				return err
			}
			if !sent {
				fresh = append(fresh, h)
			}
		}
		return nil
	})
	if err != nil || len(fresh) == 0 {
		return 0, err
	}

	sentCount := 0
	for _, c := range chans {
		var batch []hints.View
		for _, h := range fresh {
			if h.Severity >= c.MinSeverity {
				batch = append(batch, h)
			}
		}
		if len(batch) == 0 {
			continue
		}
		if err := sendBatch(ctx, d, cfg, who, c, batch); err == nil {
			sentCount++
		}
	}

	return sentCount, db.WithTx(d, func(tx *sql.Tx) error {
		for _, h := range fresh {
			if err := data.LogSent(tx, userID, h.ID); err != nil {
				return err
			}
		}
		return nil
	})
}

func sendBatch(ctx context.Context, d *sql.DB, cfg settings.Settings, who *access.Principal, c *model.NotifyChannel, batch []hints.View) error {
	var rawURL string
	err := db.WithTx(d, func(tx *sql.Tx) error {
		u, err := data.Channel(tx, c.ID)
		if err != nil || u == nil {
			return orNotFound(err)
		}
		rawURL, err = crypto.Decrypt(u.URLEnc, crypto.PurposeNotify)
		return err
	})
	if err != nil {
		return err
	}

	title := i18n.T("notify.title", who.Locale, map[string]any{"count": len(batch)})
	var lines []string
	for i, h := range batch {
		if i >= maxDigestLines {
			break
		}
		lines = append(lines, "• "+h.Title)
	}
	return outbound.Send(ctx, cfg.AppriseAPIURL, rawURL, title, strings.Join(lines, "\n"))
}

File: app/services/notify.py
"""Notifications per user: Apprise channels, quiet hours, daily/weekly digest mail.

    every minute   new hints ≥ channel level, not sent yet, outside quiet hours → push
    every 5 min    digest time reached, not sent today → HTML mail (SMTP)
"""

import logging
from dataclasses import dataclass
from datetime import date, datetime, time
from zoneinfo import ZoneInfo

from app.db.base import session_scope, utcnow
from app.db.models import NotifyChannel
from app.enums import Locale, Severity
from app.metrics import deadlines as dl
from app.outbound import notify as outbound
from app.repos import content, users
from app.repos import data as repo
from app.services import access, audit, crypto, hints, mail
from app.services.access import AccessDenied, Principal
from app.services.crypto import Purpose
from app.services.i18n import t
from app.services.mail import Row

log = logging.getLogger(__name__)

TZ = ZoneInfo("Europe/Berlin")
QUIET = "quiet"
DIGEST = "digest"
DIGEST_SENT = "digest_sent"
MAX_LINES = 10
DEADLINE_DAYS = 14
WEEKDAYS = ("mon", "tue", "wed", "thu", "fri", "sat", "sun")
LEVEL_COLOR = {Severity.INFO: "#83a598", Severity.WARN: "#fabd2f", Severity.CRITICAL: "#fb4934"}


class NotifyError(ValueError):
    pass


@dataclass
class ChannelView:
    id: int
    name: str
    hint: str
    min_severity: Severity
    enabled: bool


def _mask(url: str) -> str:
    """ntfy://ntfy.lan/topic → ntfy://…/topic (no credentials on screen)."""
    scheme, _sep, rest = url.partition("://")
    return f"{scheme}://…/{rest.rsplit('/', 1)[-1]}" if rest else scheme


def channels(who: Principal) -> list[ChannelView]:
    with session_scope() as s:
        return [
            ChannelView(c.id, c.name, _mask(crypto.decrypt(c.url_enc, Purpose.NOTIFY)), Severity(c.min_severity),
                        c.enabled)
            for c in repo.channels(s, who.user_id)
        ]


def add_channel(who: Principal, name: str, url: str, level: Severity) -> None:
    url = url.strip()
    if not outbound.valid(url):
        raise NotifyError("notify.invalid_url")
    with session_scope() as s:
        repo.add(s, NotifyChannel(user_id=who.user_id, name=name.strip() or url.split(":")[0],
                                  url_enc=crypto.encrypt(url, Purpose.NOTIFY), min_severity=int(level)))
        audit.log(s, who.user_id, "notify.channel_added", target=name)


def _own(s, who: Principal, channel_id: int) -> NotifyChannel:
    item = repo.channel(s, channel_id)
    if item is None or item.user_id != who.user_id:
        raise AccessDenied("channel")
    return item


def delete_channel(who: Principal, channel_id: int) -> None:
    with session_scope() as s:
        repo.remove(s, _own(s, who, channel_id))


def test_channel(who: Principal, channel_id: int) -> None:
    with session_scope() as s:
        url = crypto.decrypt(_own(s, who, channel_id).url_enc, Purpose.NOTIFY)
    try:
        outbound.send(url, t("notify.test_title", who.locale), t("notify.test_body", who.locale))
    except outbound.NotifyFailed as exc:
        raise NotifyError("notify.failed") from exc


# ── Preferences ──


@dataclass
class Prefs:
    quiet_from: str
    quiet_to: str
    daily: str
    weekly: str


def prefs(who: Principal) -> Prefs:
    with session_scope() as s:
        raw = dict(users.get(s, who.user_id).prefs or {})
    quiet, digest = raw.get(QUIET) or {}, raw.get(DIGEST) or {}
    return Prefs(quiet.get("from", ""), quiet.get("to", ""), digest.get("daily", ""), digest.get("weekly", ""))


def save_prefs(who: Principal, value: Prefs) -> None:
    for text in (value.quiet_from, value.quiet_to, value.daily):
        if text and not _clock(text):
            raise NotifyError("notify.bad_time")
    if value.weekly and value.weekly not in WEEKDAYS:
        raise NotifyError("notify.bad_time")

    with session_scope() as s:
        user = users.get(s, who.user_id)
        raw = dict(user.prefs or {})
        raw[QUIET] = {"from": value.quiet_from, "to": value.quiet_to}
        raw[DIGEST] = {"daily": value.daily, "weekly": value.weekly}
        user.prefs = raw


def _clock(text: str) -> time | None:
    try:
        return time.fromisoformat(text)
    except ValueError:
        return None


def quiet_now(raw: dict, now: datetime) -> bool:
    """Quiet window may cross midnight: 22:00–07:00."""
    quiet = raw.get(QUIET) or {}
    start, end = _clock(quiet.get("from", "")), _clock(quiet.get("to", ""))
    if not start or not end:
        return False
    current = now.astimezone(TZ).time()
    if start <= end:
        return start <= current < end
    return current >= start or current < end


# ── Dispatch (jobs) ──


def dispatch() -> int:
    """Push new hints to every user's channels. Returns the number of messages."""
    sent = 0
    now = utcnow()
    with session_scope() as s:
        people = [(u.id, dict(u.prefs or {})) for u in users.all_users(s) if u.is_active]

    for user_id, raw in people:
        if quiet_now(raw, now):
            continue
        sent += _dispatch_user(user_id)
    return sent


def _dispatch_user(user_id: int) -> int:
    with session_scope() as s:
        who = access.principal(s, user_id)
        chans = [(c.id, c.name, crypto.decrypt(c.url_enc, Purpose.NOTIFY), c.min_severity)
                 for c in repo.channels(s, user_id) if c.enabled]
    if who is None or not chans:
        return 0

    open_hints = hints.active(who)
    with session_scope() as s:
        new = [h for h in open_hints if not repo.was_sent(s, user_id, h.id)]
    if not new:
        return 0

    sent = 0
    for _cid, name, url, level in chans:
        batch = [h for h in new if h.severity >= level]
        if not batch:
            continue
        title = t("notify.title", who.locale, count=len(batch))
        body = "\n".join(f"• {h.title}" for h in batch[:MAX_LINES])
        try:
            outbound.send(url, title, body)
            sent += 1
        except outbound.NotifyFailed:
            log.warning("channel %s of user %s failed", name, user_id)

    with session_scope() as s:
        for hint in new:
            repo.log_sent(s, user_id, hint.id)
    return sent


def digests(now: datetime | None = None) -> int:
    """Send due digest mails (daily at HH:MM, weekly on the chosen weekday at the same time)."""
    now = (now or utcnow()).astimezone(TZ)
    sent = 0
    with session_scope() as s:
        people = [(u.id, dict(u.prefs or {})) for u in users.all_users(s) if u.is_active]

    for user_id, raw in people:
        digest = raw.get(DIGEST) or {}
        at = _clock(digest.get("daily") or "")
        if at is None or now.time() < at or raw.get(DIGEST_SENT) == now.date().isoformat():
            continue
        weekly = digest.get("weekly")
        if weekly and WEEKDAYS[now.weekday()] != weekly:
            continue
        _send_digest(user_id, now.date())
        sent += 1
    return sent


def _send_digest(user_id: int, today: date) -> None:
    with session_scope() as s:
        who = access.principal(s, user_id)
        user = users.get(s, user_id)
        prefs_ = dict(user.prefs or {})
        prefs_[DIGEST_SENT] = today.isoformat()
        user.prefs = prefs_
        settings = [dict(sp.settings or {}) for sp in content.spaces(s, list(who.spaces))]

    locale = who.locale
    rows = [Row(t(f"severity.{h.severity.name.lower()}", locale), h.title, LEVEL_COLOR[h.severity])
            for h in hints.active(who)[:MAX_LINES * 2]]
    for space_settings in settings:
        for item in dl.upcoming(space_settings, today, DEADLINE_DAYS):
            text = t(f"deadline.{item['kind']}", locale, period=item.get("period", ""), year=item.get("year", ""))
            rows.append(Row(item["due"].isoformat(), text))

    subject = t("notify.digest_subject", locale, count=len(rows))
    body = t("notify.digest_body", locale) if rows else t("hints.none", locale)
    mail.send(mail.render(who.email, Locale(locale), subject, [body], items=rows))
